File: common/src/main/java/juuxel/adorn/block/PrismarineChimneyBlock.java
package juuxel.adorn.block;

import net.minecraft.core.BlockPos;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.tags.FluidTags;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.BubbleColumnBlock;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.redstone.Orientation;
import org.jspecify.annotations.Nullable;

public class PrismarineChimneyBlock extends AbstractChimneyBlock implements BlockWithDescription {
    private static final String DESCRIPTION_KEY = "block.adorn.prismarine_chimney.description";

    public PrismarineChimneyBlock(BlockBehaviour.Properties settings) {
        super(settings);
    }

    @Override
    public String getDescriptionKey() {
        return DESCRIPTION_KEY;
    }

    @Override
    public void animateTick(BlockState state, Level world, BlockPos pos, RandomSource random) {
        if (!state.getFluidState().is(FluidTags.WATER) || state.getValue(CONNECTED)) return;

        double x = pos.getX() + 0.5;
        double y = pos.getY() + 0.9;
        double z = pos.getZ() + 0.5;

        for (int i = 0; i < 3; i++) {
            world.addAlwaysVisibleParticle(ParticleTypes.BUBBLE_COLUMN_UP, x, y, z, 0.0, 0.0, 0.0);
        }
    }

    public static final class WithColumn extends PrismarineChimneyBlock {
        private final boolean drag;

        public WithColumn(boolean drag, Properties settings) {
            super(settings);
            this.drag = drag;
        }

        public boolean getDrag() {
            return drag;
        }

        @Override
        public void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
            BubbleColumnBlock.updateColumn(Blocks.BUBBLE_COLUMN, world, pos.above(), state);
        }

        @Override
        public void onPlace(BlockState state, Level world, BlockPos pos, BlockState oldState, boolean notify) {
            world.scheduleTick(pos, this, 20);
        }

        @Override
        protected void neighborChanged(BlockState state, Level world, BlockPos pos, Block sourceBlock, @Nullable Orientation wireOrientation, boolean notify) {
            world.scheduleTick(pos, this, 20);
        }

        @Override
        public void animateTick(BlockState state, Level world, BlockPos pos, RandomSource random) {
            if (!drag) {
                super.animateTick(state, world, pos, random);
            }
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/PrismarineChimneyBlock.kt
package juuxel.adorn.block

import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.BubbleColumnBlock
import net.minecraft.particle.ParticleTypes
import net.minecraft.server.world.ServerWorld
import net.minecraft.tag.FluidTags
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import java.util.Random

open class PrismarineChimneyBlock(settings: Settings) : AbstractChimneyBlock(settings), BlockWithDescription {
    override val descriptionKey get() = AdornBlocks.PRISMARINE_CHIMNEY.translationKey + ".description"

    override fun randomDisplayTick(state: BlockState, world: World, pos: BlockPos, random: Random) {
        if (!state.fluidState.isIn(FluidTags.WATER) || state[CONNECTED]) return

        val x = pos.x + 0.5
        val y = pos.y + 0.9
        val z = pos.z + 0.5

        for (i in 1..3) {
            world.addImportantParticle(ParticleTypes.BUBBLE_COLUMN_UP, x, y, z, 0.0, 0.0, 0.0)
        }
    }

    class WithColumn(val drag: Boolean, settings: Settings) : PrismarineChimneyBlock(settings) {
        override fun scheduledTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
            BubbleColumnBlock.update(world, pos.up(), state)
        }

        override fun onBlockAdded(state: BlockState, world: World, pos: BlockPos, oldState: BlockState, moved: Boolean) {
            world.createAndScheduleBlockTick(pos, this, 20)
        }

        override fun neighborUpdate(
            state: BlockState, world: World, pos: BlockPos, neighbor: Block, neighborPos: BlockPos, moved: Boolean
        ) {
            world.createAndScheduleBlockTick(pos, this, 20)
        }

        override fun randomDisplayTick(state: BlockState, world: World, pos: BlockPos, random: Random) {
            if (!drag) {
                super.randomDisplayTick(state, world, pos, random)
            }
        }
    }
}
